### PR TITLE
refactor: refactor stub-components.js

### PR DESCRIPTION
**Pull request by lusarz**
Pull request contains changes like
- move public (exported) methods to top of `stub-components.js` file
- separate stub options validation from creation of stubs
- clean a few repeated code
- overall cleaning

### CODE REVIEW

File: packages/test-utils/src/shallow.js [hunk 1]
```diff
@@ -26,15 +26,11 @@ export default function shallow (
...
-  const stubbedComponents = createComponentStubsForAll(component)
-  const stubbedGlobalComponents = createComponentStubsForGlobals(vue)
-
   return mount(component, {
     ...options,
     components: {
-      // stubbed components are used instead of original components components
-      ...stubbedGlobalComponents,
-      ...stubbedComponents
+      ...createComponentStubsForGlobals(vue),
```
Comment (lusarz) [inline packages/test-utils/src/shallow.js:32]: `stubbedComponents` and `stubbedGlobalComponents` was unnecessary variables, I invoked functions directly

File: packages/shared/stub-components.js [hunk 2]
```diff
@@ -37,150 +72,15 @@ function getCoreProperties (component: Component): Object {
...
-
-  // Loop through extended component chains to stub all child components
-  while (extended) {
-    if (extended.components) {
-      stubComponents(extended.components, stubbedComponents)
-    }
-    extended = extended.extends
+function addToIgnoredElements (name) {
+  // ignoreElements does not exist in Vue 2.0.x
+  if (Vue.config.ignoredElements) {
```
Comment (lusarz) [inline packages/shared/stub-components.js:93]: These code was duplicated, so I moved it into function

File: packages/shared/stub-components.js [hunk 3]
```diff
@@ -37,150 +72,15 @@ function getCoreProperties (component: Component): Object {
...
-  const stubbedComponents = {}
-
-  if (component.components) {
-    stubComponents(component.components, stubbedComponents)
-  }
-
-  let extended = component.extends
-
-  // Loop through extended component chains to stub all child components
-  while (extended) {
```
Comment (lusarz) [inline packages/shared/stub-components.js:160]: Moved this logic into `createComponentStubsRecursively`

**Comment by 38elements** [comment]
This looks like only changing style.
Would you tell me the benefits of this?

**Comment by lusarz** [comment]
Yes, main purpose of refactoring is changing style and improving the design of an existing code.

I think after these changes code would be easier to maintain and it would be easier to find bugs and introduce extensions.

**Comment by 38elements** [comment]
Thank you for replying.

There is no objective basis for its benefit.

IMHO,
This is cosmetic change.
I think this change makes difficult to `git blame`. 

This makes difficult to understand code for me.
For example, the `createBlankStub` function helps to understanding code for me,
this function is removed.

**Comment by lusarz** [comment]
True, I agree with `git blame` issue. Main reason of that is that I moved public (exported) functions to top of file, what I think is good practice.

I replaced `createBlankStub` with `createStub` method, which returns 'blank' stub by default. But yeah, I agree it was no perfect solution. I updated it in [second commit](https://github.com/vuejs/vue-test-utils/pull/541/commits/25c1adc7e5cdfd800cb0cb3df14ae8d1443be94c). What do you think about that?

**Comment by 38elements** [comment]
I am sorry.
My expression may have been ambiguous.
It is important whether there is benefit or not.
IMHO,
Since this change has no benefit, I think this change is not necessary.
Other people may not think so.

**Comment by lusarz** [comment]
Reducing number of lines of code is beneficial itself

**Comment by lusarz** [comment]
I'm not yet happy with everything, so I'm gonna close this pull request for now

**Comment by eddyerburgh** [comment]
@38elements this part of the codebase is in need of refactoring, so I think a cosmetic change is fine. If we reduce the lines significantly and keep the code readable, that's a good benefit

**Comment by 38elements** [comment]
@eddyerburgh 
~~I think that a change which is consistent with the style of existing code is fine.
I think that it is fine if the readability is improved explicitly.~~

**Comment by 38elements** [comment]
@eddyerburgh 
Since you are a maintainer, 
I think that the code you are easy to maintain is correct.
Would you please do not mind my opinion.

**Pull request closed without merging**